docs/design/copy-tracking.md

```diff
@@ -10,17 +10,671 @@ results of rebasing across remote copies.
+$ jj new L M N
+```
+
+#### Propagating changes across copies?
+
+Should we propagate changes to copies? For example, if you've modified file
+`foo` and then rebase it onto a commit that copied `foo` to `bar`, should we
+apply your change to `bar` too? Mercurial does that but Git doesn't. It's
+particularly useful when a file has been split in two. For example, let's say
+you've made various changes in file `foo` and then rebase those change onto a
+commit that split `foo` into `foo1` and `foo2` (or `foo` and `bar`). If we
+propagate the changes to both files, each change will apply successfully in one
+file (assuming the changes do not overlap with the split boundary). Each change
+will have a modify/delete conflicts in the other file. Those can relatively
+easily be resolved in favor of the deleted hunk. If we do not propagate changes,
+then changes that belong in one of the files will instead only appear as
+modify/delete conflicts in the first file and you will have to manually copy
+over the changes to the copied file.
+
+Propagating changes to copies means that rebasing a commit and then rebasing it
+back is no longer a no-op even when ignoring the "same-change rule". For
+example, if your commit modifies file `foo` and you rebase that commit onto a
+commit that copied `foo` to `bar`, and then you rebase it back, the same change
+will be applied twice to `foo`. However, thanks to the same-change rule, we
+won't consider it a conflict, so maybe it actually works well in practice.
+
+A third option is to not leave it up to the user whether to propagate the
+change across the copy. We can do this by leaving the relevant paths in the
+input trees unchanged in the conflicted commit. Then we will redo the copy
+tracking process every time the commit is inspected. We can have `jj resolve`
+ask the user if they want to propagate the changes to the copy target with a
+simple yes/no question per copy target.
+
+Decision: Asking the user about propagating copies seems like the best option.
+It avoids surprises, and it makes the conflict algebra work in more cases.
+
+#### Example: Propagate changes to copied file, then rebase back
+
+```
+M foo="M"
+|
+| L copy foo->bar
+|/
+K add foo="K"
+```
+
+Let's say we rebase `M` onto `L`. Since we decided to not automatically
+propagate changes to copies, we will leave the `M+(L-K)` tree unresolved (i.e.
+without making any changes to the three trees). If the user does not resolve
+the conflict, and instead rebases `L` back onto `K`, the conflict will be
+resolved automatically per the usual conflict simplification.
+
+#### Example: Multiple copies
+
+```
+N foo="N"
+|
+| M foo="M, foo2="M2", foo3="M3"
+| |
+| L copy foo->foo2, copy foo->foo3
+|/
+K add foo="K"
+```
+
+Let's say we rebase `M` onto `N`. The changes to `foo`, `foo2`, and` foo3` will
+then all apply to `foo`, which means we get a 4-sided conflict.
+
+#### Example: Convergent renames
+
+Consider this "convergent copy/rename" scenario:
+```
```

Comment:
Minor: It's a bit too easy to miss `jj new B C` here. This will look clearer in some renderings if you use the "console" syntax highlighting here. Perhaps also add a new line before `$ jj new`?

Reply:
Done